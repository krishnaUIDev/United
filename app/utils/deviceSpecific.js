/**
 * Place for all functions that help in identifying or performing device specific actions
 */

export const isIPhone = () =>
  !!navigator.platform && /iPhone/.test(navigator.platform);

export const isAndroid = () =>
  !!navigator.platform && /Android/.test(navigator.platform);

// START Calendar management
export const isMobile = (screen.width <= 480) === true;

// @NOTE this can be applied to android too
export const shortIPhone = () => window.innerWidth <= 667;

// largest width considered for phones
export const hasPhoneWidth = () => window.innerWidth <= 812;

/**
 * For iphone or android, default to vertical,
 * otherwise consider the width (as set by isMobile)
 * @TODO discuss "isMobile logic" with Sean (screen.width does not change for devices in horizontal mode)
 */
export const isInPortraitMode = () => {
  let orientation;

  if (isIPhone()) {
    orientation = (window.orientation === 90 || window.orientation === -90) ? 'horizontal' : 'vertical';
  } else if (isAndroid()) {
    orientation = (screen.orientation.angle === 90 || screen.orientation.angle === -90) ? 'horizontal' : 'vertical';
  } else {
    return isMobile;
  }

  return orientation !== 'horizontal';
};

export const hasNumberOfMonths = () =>
  isInPortraitMode() ? {} : { numberOfMonths: shortIPhone() ? 1 : 2 };

  // START Calendar management
