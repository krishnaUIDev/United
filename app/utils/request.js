import axios from 'axios';
import { CANCEL } from 'redux-saga';

/**
 * Parses the JSON returned by a network request
 *
 * @param  {object} response A response from a network request
 *
 * @return {object}          The parsed JSON from the request
 */
function parseJSON(response) {
  const parsedResponse = response.data;
  parsedResponse.status = response.status;
  return parsedResponse;
}

/**
 * Checks if a network request came back fine, and throws an error if not
 *
 * @param  {object} response   A response from a network request
 *
 * @return {object|undefined} Returns either the response, or throws an error
 */
function checkStatus(response) {
  if (response.status >= 200 && response.status < 300) {
    return response;
  }

  const error = new Error(response.statusText);
  error.response = response;
  delete error.config;
  delete error.request;
  throw error;
}

/**
 * Requests a URL, returning a promise
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export default function request(url, options = {}) {
  const cancelTokenSource = axios.CancelToken.source();
  // console.log('cancelTokenSource.token', cancelTokenSource.token);
  const promise = axios(url, Object.assign(options, { cancelToken: cancelTokenSource.token }))
    .then(checkStatus)
    .catch((error) => {
      // console.log('axios catch', error);
      /* istanbul ignore next */
      const errorObj = new Error((error && error.response && error.response.statusText) || error.message);
      errorObj.response = error && error.response;
      errorObj.code = error && error.code; // ECONNABORTED
      delete errorObj.config;
      delete errorObj.request;
      // console.log('axios catch %o', errorObj, errorObj.code);
      throw errorObj;
    })
    .then(parseJSON);

  /* istanbul ignore next */
  promise[CANCEL] = () => {
    // cancel XHR request, called by redux-saga when a saga gets cancelled
    cancelTokenSource.cancel('Operation canceled.');
  };

  return promise;
}
