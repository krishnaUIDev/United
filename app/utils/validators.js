export const passwordPatternAllowEmpty = /^$|[ -~]{8,}$/;

/**
 * @param {string} val
 * @param {string} globalMPusername
 */
export const isSubmittableInput = (val) => {
  if (!val) return false;

  const loginPattern = /^(?=(?:.{8}|.{11})$)[a-zA-Z0-9]{8,11}$/;
  return (val.indexOf('**') >= 0) || (isString(val) && (loginPattern.test(val)));
};

/**
 * @param {string} val
 * @return {boolean} true is input is empty or submittable
 */
export const isValidMPnumber = (val, globalMPusername) =>
  isEmptyString(val) || isSubmittableInput(val, globalMPusername);

/**
 * @param {string} val
 */
export const isSubmittablePassword = (val) => {
  if (!val) return false;

  // TODO: change this check once the enter key rrf is fixed. Link: https://github.com/davidkpiano/react-redux-form/issues/1011
  const passwordPattern = /^[ -~]{8,}$/;
  return isString(val) && (passwordPattern.test(val));
};

/**
 * @param {string} val
 */
export const isValidPassword = (val) =>
  isEmptyString(val) || isSubmittablePassword(val);

// START Helpers
export function isString(val) {
  return (typeof (val) === 'string');
}
export function isEmptyString(val) {
  return isString(val) && !val;
}
// END Helpers
