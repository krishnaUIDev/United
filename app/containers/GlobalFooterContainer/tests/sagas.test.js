import { cancel, take, put, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';
import { LOCATION_CHANGE } from 'react-router-redux';

import serializeError from 'utils/serializeError';

import {
  LOAD_FOOTER_DATA,
} from '../constants';

import {
  footerLoaded,
  footerLoadingError,
} from '../actions';

import {
  getFooterData,
  footerData,
} from '../sagas';

import sample from '../samples/sample.json';

describe('getFooterData Saga', () => {
  let getFooterDataGenerator;
  beforeEach(() => {
    getFooterDataGenerator = getFooterData();

    const selectDescriptor = getFooterDataGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor2 = getFooterDataGenerator.next().value;
    expect(callDescriptor2).toMatchSnapshot();
    const callDescriptor3 = getFooterDataGenerator.next({ hash: '1234abc', expiresAt: '2020-11-02T15:43:42.0000000Z' }).value;
    expect(callDescriptor3).toMatchSnapshot();
  });

  it('should dispatch the footerLoaded action if it requests the data successfully', () => {
    const response = sample; // Sample api response with multiple results
    const putDescriptor = getFooterDataGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(footerLoaded(response.data.headerFooter)));
  });

  it('should call the footerLoadingError action if the response errors with status 403', () => {
    const error = ({
      response: {
        status: 403,
      },
    });
    const response = new Error(error);
    getFooterDataGenerator.throw(response);
    const putDescriptor = getFooterDataGenerator.next().value;
    expect(putDescriptor).toEqual(put(footerLoadingError(serializeError(response))));
  });
});

describe('carouselSlidesData Saga', () => {
  const footerDataSaga = footerData();
  const mockedTask = createMockTask();

  it('should start task to watch for LOAD_FOOTER_DATA action', () => {
    const takeLatestDescriptor = footerDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_FOOTER_DATA, getFooterData));
  });

  it('should yield until LOCATION_CHANGE action', () => {
    const takeDescriptor = footerDataSaga.next(mockedTask).value;
    expect(takeDescriptor).toEqual(take(LOCATION_CHANGE));
  });

  it('should cancel the forked task when LOCATION_CHANGE happens', () => {
    const cancelDescriptor = footerDataSaga.next().value;
    expect(cancelDescriptor).toEqual(cancel(mockedTask));
  });
});
