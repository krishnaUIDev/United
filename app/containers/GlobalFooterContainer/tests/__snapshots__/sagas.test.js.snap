// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFooterData Saga should call the footerLoadingError action if the response errors with status 403 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getFooterData Saga should call the footerLoadingError action if the response errors with status 403 2`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getFooterData Saga should call the footerLoadingError action if the response errors with status 403 3`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/home/headerFooter",
      Object {
        "headers": Object {
          "Accept": "application/json",
          "Accept-Language": "en-US",
          "Authorization": "bearer 1234abc",
          "Content-Type": "application/json",
        },
        "timeout": 20000,
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getFooterData Saga should dispatch the footerLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getFooterData Saga should dispatch the footerLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getFooterData Saga should dispatch the footerLoaded action if it requests the data successfully 3`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/home/headerFooter",
      Object {
        "headers": Object {
          "Accept": "application/json",
          "Accept-Language": "en-US",
          "Authorization": "bearer 1234abc",
          "Content-Type": "application/json",
        },
        "timeout": 20000,
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
